### ISSUE TITLE
Performance of validation of JSON objects with very long lists of JSON objects within is very, very slow.

### ISSUE THREAD

**Issue by germanblanco** [for-GET/jesse#115]
It seems that validating an object with a list of 1000 objects with uniqueItems property takes time in the order of one second to complete.

This may be verified by creating an object such as:
```
{
  "test": [
    {"key1": "1", "key2": "1", "key3": "1", },
    {"key1": "2", "key2": "2", "key3": "3", },
    ...
    {"key1": "1000", "key2": "1000", "key3": "1000", },
  ]
}
```

and validating it against an schema such as:

```
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "definitions": {
    "dataType": {
      "type": "object",
      "properties": {
          "key1": "string",
          "key2": "string",
          "key3": "string"
      }
    }
  },
  "type": "object",
  "properties": {
    "test": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "$ref": "#/definitions/dataType"
      }
    }
  }
}
```

**Comment by germanblanco** [comment]
One solution to this is already verified for forked project https://github.com/kobil-systems/jessy

The solution is here: https://github.com/kobil-systems/jessy/pull/7

I have prepared already a patch to this project and would love to have the opportunity to submit a pull request with this proposal.

**Comment by seriyps** [comment]
Hi,

do you have any idea if the performance is not great because of nested loop

```erlang
lists:foldl( fun(Item, RestItems) ->
                     lists:foreach(...
```

or because of slow comparison function `jesse_validator_draft4:is_equal/2`?

I have at least one problem with the implementation you have in your fork - it uses `sets` module with `{version, 2}`, which is not available on OTP versions prior to, I think, 24. Jesse currently supports OTP versions starting from 19. I would suggest using `lists:usort` instead.

**Comment by seriyps** [comment]
The current nested loop seems to be slightly optimized quadratic complexity (it needs 45 calls to `is_equal` for list of 10 elements, 1225 for 50 elements and  4950 for 100 elements), which is indeed far from optimal =(

Are you sure your `jesse_lib:normalize_and_sort` is equivalent to `is_equal`? I can try to write a property-based test...

**Comment by germanblanco** [comment]
> Hi,
> 
> do you have any idea if the performance is not great because of nested loop
> 
> ```erlang
> lists:foldl( fun(Item, RestItems) ->
>                      lists:foreach(...
> ```
> 
> or because of slow comparison function `jesse_validator_draft4:is_equal/2`?
> 
> I have at least one problem with the implementation you have in your fork - it uses `sets` module with `{version, 2}`, which is not available on OTP versions prior to, I think, 24. Jesse currently supports OTP versions starting from 19. I would suggest using `lists:usort` instead.

My impression is that the main reason is the nested loop.  If I am not wrong that means this search is O(n^2), while creating a set, which should be creating a hash for each element and then comparing, is O(n).  The observation is that the time that it takes grows more than proportionally with the number of elements in the list.

**Comment by germanblanco** [comment]
> 
> I have at least one problem with the implementation you have in your fork - it uses `sets` module with `{version, 2}`, which is not available on OTP versions prior to, I think, 24. Jesse currently supports OTP versions starting from 19. I would suggest using `lists:usort` instead.

Thank you very much for the suggestion!  I have no problem at all in changing the implementation, that is in fact one of the reasons to propose this change here.  Is it enough if I create a new pull request in the child project using lists:usort after testing it? Or do you want to see the new proposals in any other way? Would it be possible for me to get permissions to create a pull request here and discuss the implementation there?

**Comment by germanblanco** [comment]
> The current nested loop seems to be slightly optimized quadratic complexity - `0.5 * x^2 - 0.5 * x`, which is indeed far from optimal =(
> 
> list len	num calls to is_equal
> 10	45
> 50	1225
> 100	4950
> 500	124750
> 1000	499500
> http://tryerl.seriyps.ru/#id=fbb4
> 
> Are you sure your `jesse_lib:normalize_and_sort` is equivalent to `is_equal`? I can try to write a property-based test...

I am going to try that is_equal thing.  I have to say that my knowledge of the library is very little at the moment, so there might be many options that I just didn't considered for lack of knowledge. It was just a matter of solving this problem with what I could do at the moment, since it was really dangerous as it was.

**Comment by germanblanco** [comment]
> Hi,
> 
> do you have any idea if the performance is not great because of nested loop
> 
> ```erlang
> lists:foldl( fun(Item, RestItems) ->
>                      lists:foreach(...
> ```
> 
> or because of slow comparison function `jesse_validator_draft4:is_equal/2`?
> 
> I have at least one problem with the implementation you have in your fork - it uses `sets` module with `{version, 2}`, which is not available on OTP versions prior to, I think, 24. Jesse currently supports OTP versions starting from 19. I would suggest using `lists:usort` instead.

After looking a bit into this, my proposal is to just check the OTP version and use `{version, 2}` only when available.  This is anyway an optimization that, as far as I know, it only matters in terms of memory used.  The improvement in speed is there with and without this parameter, but I would prefer to keep it for OTP versions 24 and higher. My idea is to do this with a macro, such as this one:

```
-ifdef(OTP_RELEASE).
  -if(?OTP_RELEASE >= 24). 
  %% OTP 24 or higher
    -define(SET_FROM_LIST(List), sets:from_list(List, [{version, 2}])).
  -else.
  %% OTP 23 or lower.
    -define(SET_FROM_LIST(List), sets:from_list(List)).
  -endif.
-endif.
```

**Comment by germanblanco** [comment]
> 
> Are you sure your `jesse_lib:normalize_and_sort` is equivalent to `is_equal`? I can try to write a property-based test...

I am working now in that property-based test.  Thank you very much for this idea.  If you have a suggestion for a "Generator" for lists of JSON objects, that would be awesome.

**Comment by seriyps** [comment]
>  If you have a suggestion for a "Generator" for lists of JSON objects, that would be awesome.

I actually have https://gist.github.com/seriyps/ffe0f5f5ba270fc4dd395208718e84d3

Maybe not perfect and would need to be wrapped to `proper_types:list`

**Comment by germanblanco** [comment]
This looks awesome!  Thank you very much, I hope to have something working in a day or two.

**Comment by germanblanco** [comment]
Hello again @seriyps !

I would like to propose the changes in this branch: https://github.com/kobil-systems/jessy/pull/8/files

Please tell me if this looks ok or any change in direction is needed.

The ideas behind the changes are the following:
- is_equal is moved to jesse_lib and a parameter is added so that it behaves different for draft 3 and for draft 4. This saves from repeating many lines of code that are equal and makes it much more easier to spot the change between the two behaviors.
- A module for property based testing is added. The intention is only to verify that "normalization" produces objects that are equal in terms of the is_equal function.  Maybe this is not the best way to check if normalization works, please comment.
- Normalization was also fixed in order to take into account the difference between draft 3 and draft 4 comparisons.
- The Erlang sets version {version, 2} is only used if running on OTP 24 or higher.

Please tell me what you think about all this.

**Comment by seriyps** [comment]
@germanblanco I added one coment regarding big integers in your `jessy` PR. I think it would be a bit easier if you will open a PR towards this repository, so we can continue the discussion there.

> is_equal is moved to jesse_lib and a parameter is added so that it behaves differently for draft 3 and for draft 4. This saves from repeating many lines of code that are equal and makes it much more easier to spot the change between the two behaviors.

I see in your PR you made v3 and v4 versions of `es_equal` to behave the same way even though there was a difference in how they compare integers with floats. I might need to check the specks to see if it's ok.

> A module for property based testing is added. The intention is only to verify that "normalization" produces objects that are equal in terms of the is_equal function. Maybe this is not the best way to check if normalization works, please comment.

:+1:
not yet sure what other tests can be added (maybe the one withbig integers I mentioned in other PR), but I think it is already a great start!

> During normalization, JSON objects are tagged with the "struct" atom in order to be able to tell the difference between an empty object and an empty list or set

Wouldn't it be easier to convert them to maps instead?

> The Erlang sets version {version, 2} is only used if running on OTP 24 or higher.

:+1:

**Comment by germanblanco** [comment]
Hello!
please take a look at the pull request, I have tried to address all the issues you mentioned.
It seems I need your approval so that CI checks this pull request, please consider this.
Thanks a lot again for all the great suggestions to this change.

**Comment by seriyps** [comment]
Fixed by #116
